packages/core/src/sync-realtime/service.ts

```diff
@@ -486,160 +374,393 @@ export class RealtimeSyncService extends Emittery<RealtimeSyncEvents> {
+    const localLogs = this.logs.filter(
+      (log) => log.blockNumber <= newFinalizedBlockNumber,
+    );
+
+    /**
+     * Common ancestor is the block directly before the logs diverge.
+     * If the divergence occurred at index 0, check for deep re-org.
+     */
+    const handleReorg = async (nonMatchingIndex: number) => {
+      if (nonMatchingIndex === 0) {
+        this.blocks = [];
+        this.logs = [];
+
+        const hasDeepReorg = await this.detectDeepReorg(latestBlockNumber);
+
+        if (hasDeepReorg) return;
+
+        await this.syncStore.deleteRealtimeData({
+          chainId: this.network.chainId,
+          fromBlock: BigInt(this.finalizedBlock.number),
+        });
+
+        this.emit("shallowReorg", {
+          blockTimestamp: this.finalizedBlock.timestamp,
+          chainId: this.network.chainId,
+          blockNumber: this.finalizedBlock.number,
         });
+      } else {
+        const ancestorBlockHash = localLogs[nonMatchingIndex - 1].blockHash;
+        const commonAncestor = this.blocks.find(
+          (block) => block.hash === ancestorBlockHash,
+        )!;
 
-        // Remove all non-canonical blocks from the local chain.
         this.blocks = this.blocks.filter(
-          (block) => block.number <= commonAncestorBlock.number,
+          (block) => block.number <= commonAncestor.number,
+        );
+        this.logs = this.logs.filter(
+          (log) => log.blockNumber <= commonAncestor.number,
         );
 
         await this.syncStore.deleteRealtimeData({
           chainId: this.network.chainId,
-          fromBlock: BigInt(commonAncestorBlock.number),
+          fromBlock: BigInt(commonAncestor.number),
         });
 
-        // Clear the queue of all blocks (some might be from the non-canonical chain).
-        // TODO: Figure out if this is indeed required by some edge case.
-        this.queue.clear();
-
-        // Add blocks from the canonical chain (they've already been fetched).
-        for (const block of canonicalBlocksWithTransactions) {
-          const priority = Number.MAX_SAFE_INTEGER - hexToNumber(block.number);
-          this.queue.addTask({ block }, { priority });
-        }
-
-        // Also add a new latest block, so we don't have to wait for the next poll to
-        // start fetching any newer blocks on the canonical chain.
-        await this.addNewLatestBlock();
         this.emit("shallowReorg", {
-          blockTimestamp: commonAncestorBlock.timestamp,
+          blockTimestamp: commonAncestor.timestamp,
           chainId: this.network.chainId,
-          blockNumber: commonAncestorBlock.number,
+          blockNumber: commonAncestor.number,
         });
+      }
+    };
+
+    let i = 0;
+    for (; i < localLogs.length && i < matchedLogs.length; i++) {
+      const lightMatchedLog = realtimeLogToLightLog(matchedLogs[i]);
+      if (lightMatchedLog.blockHash !== localLogs[i].blockHash) {
+        handleReorg(i);
```

Comment:
If our local chain has multiple now-invalid blocks, won't this run multiple times (unnecessarily)?

Reply:
Doesn't it return right after this line?